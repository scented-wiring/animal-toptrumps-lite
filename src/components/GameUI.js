import "../styles/GameUI.css";
import { Link } from "react-router-dom";

const GameUI = ({
  playerTurn,
  playField,
  win,
  lostCard,
  clearWin,
  cardHighField,
  gameWinner,
  tieCards,
  tieCardsLength,
  noCards,
}) => {
  let status, message;

  if (noCards) {
    //if no cards in deck
    status = "No cards found";
    message = 'Go to "Create Cards" from the menu to add cards to your deck.';
  } else if (gameWinner) {
    //if game is over
    status = `${gameWinner} wins the game!`;
    message = "GAME OVER";
  } else if (win === "Tie") {
    //if action results in tie
    status = `${win}!`;
    message = `Top player and computer cards added to tie deck.`;
  } else if (win) {
    //if action results in win
    status = `${win} wins!`;
    message = `${lostCard}${
      //^^lists name of won top card^^
      tieCardsLength > 0 //lists names of won tie cards (if any)
        ? `,${tieCards.map((card) => {
            if (tieCards.length - 1 === tieCards.indexOf(card)) {
              return ` and ${card.name}`;
            } else {
              return ` ${card.name}`;
            }
          })} `
        : " "
    }
    added to ${win.toLowerCase()}'s deck.`;
  } else {
    if (playerTurn) {
      //if player turn
      status = "Player's turn";
      message = "Select a field to play against your opponent:";
    } else {
      //if computer turn
      status = "Computer's turn";
      message = `Computer plays ${cardHighField}...`;
    }
  }

  return (
    <div className="game-ui">
      <div className="status">{status}</div>
      {!win && !gameWinner && tieCardsLength > 0 && (
        <div className="tie">{tieCardsLength} tie cards up for grabs!</div>
      )}
      <div className="message">{message}</div>
      {gameWinner && (
        <div className="buttons">
          <button onClick={() => window.location.reload()}>Play again</button>
          <Link to="/">
            <button>Back</button>
          </Link>
        </div>
      )}
      {win && (
        <button className="clearAlert" onClick={clearWin}>
          OK
        </button>
      )}
      {playerTurn && !noCards && !win && !gameWinner && (
        <div id="fields">
          <div className="field" onClick={() => playField("cool")}>
            Cool
          </div>
          <div className="field" onClick={() => playField("largeness")}>
            Largeness
          </div>
          <div className="field" onClick={() => playField("handsome")}>
            Handsome
          </div>
        </div>
      )}
      {!playerTurn && !win && !gameWinner && (
        <button className="clearAlert" onClick={() => playField(cardHighField)}>
          OK
        </button>
      )}
    </div>
  );
};

export default GameUI;
