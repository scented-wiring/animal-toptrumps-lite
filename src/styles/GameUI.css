.game-ui {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  text-align: center;
  width: 300px;
}

.status {
  font-family: "Staatliches", sans-serif;
  text-decoration: underline;
  font-size: 40px;
  color: white;
  text-shadow: 3px 3px black;
}

.message {
  align-self: center;
  font-size: 30px;
  width: 300px;
}

.field {
  font-family: "Staatliches", sans-serif;
  font-size: 50px;
  transition: all 0.4s ease 0s;
  margin: 10px 0;
}

.field:hover {
  color: red;
  cursor: pointer;
  transition: all 0.4s ease 0s;
  text-shadow: 3px 3px black;
}

.tie {
  font-size: 30px;
  font-family: "Staatliches", sans-serif;
}

.game-ui .buttons {
  display: flex;
  flex-direction: column;
}

.game-ui .buttons button {
  margin-bottom: 15px;
}

@media only screen and (max-width: 1005px) {
  .game-ui {
    width: 250px;
  }

  .status {
    font-size: 40px;
  }

  .message {
    font-size: 18px;
    width: 220px;
  }

  .field {
    font-family: "Staatliches", sans-serif;
    font-size: 40px;
  }

  .field:hover {
    text-shadow: 2px 2px black;
  }

  .tie {
    font-size: 25px;
  }

  .game button {
    font-size: 25px;
    margin: 0 8px;
    height: 40px;
    width: 100px;
  }
}

@media only screen and (max-width: 690px) {
  .game-ui {
    justify-content: space-evenly;
    margin: 0 10px;
    width: auto;
  }

  .status {
    font-size: 30px;
  }

  .tie {
    font-size: 20px;
  }

  .message {
    width: auto;
  }

  #fields {
    display: flex;
    align-self: center;
  }

  .field {
    font-size: 25px;
    margin: 0 10px;
  }

  .game-ui .buttons button {
    margin-bottom: 8px;
  }
}

@media only screen and (max-width: 350px) {
  .game-ui {
    width: 270px;
  }

  .status {
    font-size: 22px;
  }

  .tie {
    font-size: 17px;
  }

  .message {
    font-size: 15px;
  }

  .field {
    font-size: 22px;
    margin: 0 5px;
  }

  .game-ui .buttons button {
    height: 27px;
    font-size: 13px;
    margin-bottom: 8px;
  }
}
