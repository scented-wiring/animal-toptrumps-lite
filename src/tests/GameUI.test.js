import GameUI from "../components/GameUI";
import { render, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

describe("GameUI component", () => {
  test("renders correctly", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <GameUI />
      </BrowserRouter>
    );
    expect(asFragment).toMatchSnapshot();
  });

  test("renders message when game is over", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI gameWinner={"Test"} />
      </BrowserRouter>
    );
    expect(getByText("Test wins the game!")).toBeTruthy();
    expect(getByText("GAME OVER")).toBeTruthy();
  });

  test("renders message on tie", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI win={"Tie"} />
      </BrowserRouter>
    );
    expect(getByText("Tie!")).toBeTruthy();
    expect(
      getByText("Top player and computer cards added to tie deck.")
    ).toBeTruthy();
  });

  test("renders 'Player's turn' if playerTurn is true", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI playerTurn={true} />
      </BrowserRouter>
    );
    expect(getByText("Player's turn")).toBeTruthy();
  });

  test("renders 'Computer's turn' if playerTurn is false", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI playerTurn={false} />
      </BrowserRouter>
    );
    expect(getByText("Computer's turn")).toBeTruthy();
  });

  test("renders the names of the round winner and won card", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI win={"Test"} lostCard={"Test card"} />
      </BrowserRouter>
    );
    expect(getByText("Test wins!")).toBeTruthy();
    expect(getByText("Test card added to test's deck.")).toBeTruthy();
  });

  test("renders the names of all won cards after a tie break", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI
          win={"Test"}
          lostCard={"Test card"}
          tieCardsLength={2}
          tieCards={[{ name: "Dog" }, { name: "Cat" }]}
        />
      </BrowserRouter>
    );
    expect(getByText("Test wins!")).toBeTruthy();
    expect(
      getByText("Test card, Dog, and Cat added to test's deck.")
    ).toBeTruthy();
  });

  test("renders a play again and back button if game is over", () => {
    const { getAllByRole } = render(
      <BrowserRouter>
        <GameUI gameWinner={"Test"} />
      </BrowserRouter>
    );

    expect(getAllByRole("button")).toHaveLength(2);
  });

  test("renders a clear message button after round win", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI win={"Test"} />
      </BrowserRouter>
    );
    expect(getByText("OK")).toBeTruthy();
  });

  test("renders a list of fields to select on player's turn", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI playerTurn={true} win={false} gameWinner={false} />
      </BrowserRouter>
    );
    expect(getByText("Player's turn")).toBeTruthy();
    expect(getByText("Cool")).toBeTruthy();
    expect(getByText("Largeness")).toBeTruthy();
    expect(getByText("Handsome")).toBeTruthy();
  });

  test("renders a message showing computer's selected field on computer turn", () => {
    const { getByText } = render(
      <BrowserRouter>
        <GameUI
          playerTurn={false}
          win={false}
          gameWinner={false}
          cardHighField={"cool"}
        />
      </BrowserRouter>
    );
    expect(getByText("Computer plays cool...")).toBeTruthy();
  });
});
