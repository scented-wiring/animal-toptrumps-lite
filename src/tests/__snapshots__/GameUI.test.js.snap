// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameUI component renders correctly 1`] = `[Function]`;
